src/workerd/util/symbolizer.c++

```diff
@@ -0,0 +1,133 @@
+#include <cstdint>
+#include <errno.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <sys/wait.h>
+#include <unistd.h>
+#include <kj/common.h>
+#include <kj/debug.h>
+#include <kj/exception.h>
+#include "dlfcn.h"
+
+// Stack trace symbolizer. To use link this source to the binary.
+// Current implementation shells out to $LLVM_SYMBOLIZER if it is defined.
+
+namespace kj {
+
+struct Subprocess {
+  // A simple subprocess wrapper with in/out pipes.
+
+  static kj::Maybe<Subprocess> exec(kj::StringPtr cmd) {
+    // Execute command with a shell.
+    // Since this is used during error handling we do not to try to free resources in
+    // case of errors.
+
+    int in[2]; // process stdin pipe
+    int out[2]; // process stdout pipe
+
+    if (pipe(in)) {
+      KJ_LOG(ERROR, "can't allocate in pipe", strerror(errno));
+      return nullptr;
+    }
+    if (pipe(out)){
+      KJ_LOG(ERROR, "can't allocate out pipe", strerror(errno));
+      return nullptr;
+    }
+
+    auto pid = fork();
+    if (pid > 0) {
+      // parent
+      close(in[0]);
+      close(out[1]);
+      return Subprocess({ .pid = pid, .in = in[1], .out = out[0] });
+    } else {
+      // child
+      close(in[1]);
+      close(out[0]);
+
+      // redirect stdin
+      close(0);
+      if(dup(in[0]) < 0) {
+        _exit(2 * errno);
+      }
+
+      // redirect stdout
+      close(1);
+      if(dup(out[1]) < 0) {
+        _exit(3 * errno);
+      }
+
+      execl("/bin/sh", "sh", "-c", cmd.cStr(), nullptr);
+      _exit(1);
+    }
+  }
+
+  int closeAndWait() {
+    close(in);
+    close(out);
+    int status;
+    waitpid(pid, &status, 0);
+    return status;
+  }
+
+  int pid;
+  int in;
+  int out;
+};
+
+String stringifyStackTrace(ArrayPtr<void* const> trace) {
+  const char* llvmSymbolizer = getenv("LLVM_SYMBOLIZER");
```

Comment:
I tried using this but it appears bazel hides all environment variables from test programs, so it doesn't work unless I run the test directly. What's the right way to make it work when the test runs under bazel?

Reply:
(wrong chain)

it is in the top comment: `-c dbg --test_env=LLVM_SYMBOLIZER=llvm-symbolizer-15`

we can always make things default later

Reply:
Duh, sorry, I can't read.